#### This is the mumame output analyzer script ####

############# This first part of the script is to be user edited.
############# Please fill in file names etc. here, but avoid messing with the rest of the code unless you know what you are doing!

############# Enter the name of the output file from Mumame here (usually this ends either with ".table.txt" or ".per-mutation.txt"
filename   = "input.per-mutation.txt"

############# Enter your desired start for the filenames generated by this R script, the main statistics will be in a file ending with "mumame_stats.txt"
outputname = "output"

############# Enter your test setup below.
############# This can be an exposure gradient with increasing numbers, or a categorical testing variable (e.g. corresponding to treated/untreated).
############# In the latter case, this should be represented by zeros (0) for the controls and ones (1) for the treated.
############# In this version, only two categories can be tested.
############# All variables should be seprated by commas, e.g. c(0,0,0,1,1,1) for three controls and three treated.
############# Make sure that the sample order is the same as for the columns in the Mumame output file!
testVariable = c(0,0,0,1,1,1)

############# Enter the name of the tested variable here
testVariableName = "Treatment"


#### Now, some license information:
  # MuMaMe - Mutation Mapping in Metagenomes
  #   Copyright (C) 2018-2019 Johan Bengtsson-Palme & Viktor Jonsson

  #  Redistribution and use in source and binary forms, with or without
  #  modification, are permitted provided that the following conditions are met:
  #      * Redistributions of source code must retain the above copyright
  #        notice, this list of conditions and the following disclaimer.
  #      * Redistributions in binary form must reproduce the above copyright
  #        notice, this list of conditions and the following disclaimer in the
  #        documentation and/or other materials provided with the distribution.
  #      * Neither the name of University of Gothenburg, nor the
  #        names of its contributors may be used to endorse or promote products
  #        derived from this software without specific prior written permission.

  #  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 'AS IS' AND
  #  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
  #  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
  #  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE FOR ANY
  #  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  #  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  #  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  #  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  #  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
  #  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


#### Do not modify the code below unless you know what you are doing! ####

mutTab = read.table(filename, header = T, sep = "\t", quote = "", row.names = 1)

mut = c(rep(1,(ncol(mutTab)/2)),rep(0,(ncol(mutTab)/2)))
testVar = c(testVariable,testVariable)
libSize = as.double(mutTab[nrow(mutTab),])

output = matrix(c("Ef-mut+test", "Ef-ratio", "p.anova", "p.ratio"), 1, 4)

pdf(paste(outputname,"plots.pdf",sep = "."), height = 3.5, width = 3.2)

for (l in (1:(nrow(mutTab)-1))) {
mutvariants = mutTab[l,1:(ncol(mutTab)/2)]
wtvariants = mutTab[l,(ncol(mutTab)/2+1):(ncol(mutTab))]
variants = as.double(mutTab[l,])+1
proportions = variants / libSize
mutprop = variants[1:(ncol(mutTab)/2)] / (variants[1:(ncol(mutTab)/2)] + variants[(ncol(mutTab)/2+1):(ncol(mutTab))])

mutLM = glm(variants ~ mut * testVar + offset(log(libSize)), family = quasipoisson())

mutLM_null = glm(variants ~ mut + testVar + offset(log(libSize)), family = quasipoisson())

mutANOVA = anova(mutLM, mutLM_null, test = "F")

mutcoef = summary(mutLM)$coefficients

mutpropLM = glm(mutprop ~ testVar[1:(ncol(mutTab)/2)], weights = libSize[1:(ncol(mutTab)/2)])

mutpropcoef = summary(mutpropLM)$coefficients

output = rbind(output, c(mutcoef[4,1],mutpropcoef[2,1],mutANOVA$Pr[2],mutpropcoef[2,4]))

row.names(output)[l+1] = row.names(mutTab)[l]


mutname = strsplit(row.names(mutTab)[l],":")[[1]][1]
mutdesc = strsplit(row.names(mutTab)[l],":")[[1]][2]

plevel = ""
if (mutpropcoef[2,4] <= 0.05) {
plevel = "*"
}
if (mutpropcoef[2,4] <= 0.01) {
plevel = "**"
}
if (mutpropcoef[2,4] <= 0.001) {
plevel = "***"
}

if (sort(unique(testVariable)) == c(0,1)) {
boxplot(t(mutvariants/(wtvariants+mutvariants))[testVariable == 0],t(mutvariants/(wtvariants+mutvariants))[testVariable == 1], main = paste(mutname,plevel), sub = mutdesc, xlab = testVariableName, ylab = "Proportion mutated over wildtype", cex.axis = 0.7, cex.lab = 0.7, cex.sub = 0.5, names = c("Control", "Test"), las = 1)
boxplot(t(mutvariants/libSize)[testVariable == 0],t(mutvariants/libSize)[testVariable == 1], main = paste(mutname,plevel), sub = mutdesc, xlab = testVariableName, ylab = "Proportion of reads, mutated", cex.axis = 0.7, cex.lab = 0.7, cex.sub = 0.5, names = c("Control", "Test"), las = 1)
boxplot(t(wtvariants/libSize)[testVariable == 0],t(wtvariants/libSize)[testVariable == 1], main = paste(mutname,plevel), sub = mutdesc, xlab = testVariableName, ylab = "Proportion of reads, wildtype", cex.axis = 0.7, cex.lab = 0.7, cex.sub = 0.5, names = c("Control", "Test"), las = 1)

} else {

plot(testVariable,mutvariants/(wtvariants+mutvariants), cex = 0.7, pch = "*", main = paste(mutname,plevel), sub = mutdesc, xlab = testVariableName, ylab = "Proportion mutated over wildtype", cex.axis = 0.7, cex.lab = 0.7, cex.sub = 0.5, las = 1)
plot(testVariable,mutvariants/libSize, col = "red", pch = "*", cex = 0.7, ylim = c(0,max(variants/libSize)), main = paste(mutname,plevel), sub = mutdesc, xlab = testVariableName, ylab = "Proportion of reads", cex.axis = 0.7, cex.lab = 0.7, cex.sub = 0.5, las = 1)
points(testVariable,wtvariants/libSize, col = "blue", pch = "o", cex = 0.7)
}

}

dev.off()

write.table(output, paste(outputname,"mumame_stats.txt",sep = "."), quote = FALSE, sep = "\t", row.names = TRUE, col.names = TRUE)
